pkgs/servers/http/apache-modules/mod_csr/default.nix

```diff
@@ -0,0 +1,23 @@
+{ stdenv, fetchurl, gnused, coreutils, pkgconfig, apacheHttpd, openssl, openldap, apr, aprutil, mod_ca }:
+
+stdenv.mkDerivation rec {
+ name = "mod_csr";
+
+ meta = with stdenv.lib; {
+    description = "RedWax CA service module to handle Certificate Signing Requests.";
+    version = "0.2.1";
+
+    inherit (mod_ca.meta) license platforms maintainers homepage;
+ };
+ 
+ src = fetchurl {
+   url = "${mod_ca.baseurl}${name}-${meta.version}${mod_ca.suffix}";
+   sha256 = "01sdvv07kchdd6ssrmd2cbhj50qh2ibp5g5h6jy1jqbzp0b3j9ja";
+ };
+ buildInputs = [ mod_ca gnused coreutils pkgconfig apacheHttpd apr aprutil openssl openldap ];
+ preBuild = "cp ${./openssl_setter_compat.h} openssl_setter_compat.h";
```

Comment:
Where does this file come from? Why is it needed? What does it do? Should you submit a PR upstream to have this fixed in the source?

Reply:
That PR is working its way through OpenSSL - basically this is a set of forgotten setters/getters. But introducing them introduced another discussion about 1) if the original API is right. 2) the realisation that a lot more were forgotten and 3) concerns that introducing the setters for all of them will perhaps mean an API breaking change about how X509_ALGO* memory is handled - so this will take more than just a few months to propagate through.   And when it does - we have about 9 releases in the 1.x series that have this forgotten call.

Reply:
Sounds like a valid reason to include it here I guess. Who wrote that file? Can we find it online somewhere? Is any other distro using it? Every time we can fetch a file instead of keeping it in tree that is a win.

And documentation is key. We'll want all the facts documented, link to an upstream discussion/PR, notes as to when the file can be removed, etc... If you disappear for any reason someone else needs to track this.

Reply:
I've put the URL of the pull request in; along with the likely version by which this file can be removed and some explanation as comments.

If that is unacceptable - the other option is to wait for version 0.2.2 of RedWax - which will have moved these files into the distribution (until August this was not needed - but as of November  - quite a few distributions have the broken openssl versions as their base).

Reply:
> quite a few distributions have the broken openssl versions as their base

Do you mean the switch from `openssl` `1.0.x` to `1.1.x`? We have also switched to `1.1.x` by default, but continue to ship `1.0.x` as well. Just trying to clarify what you mean by "broken".

Reply:
I've put some comments in the default.nix file -- but basically after -1.0.2t getters were introduced and part of the openssl struct's became private. 

Most of this quite esoteric - except for the algo bits around a Certificate Signing Request (which you often need to set as the CSR may request defaults ones not acceptale to the policy of the CA) -- so a fair bit of PKI code cannot be compiled against openssl post 1.0.2t (such as  openssl-1.1.0l -to openssl-1.1.0c).

I'll put the pull request for the small fix in openssl in -- though I guess that it may linger & a bigger setter replacement will be needed (and that would also repair the open-ssh patches for x509).